cli/cmd/inject_util.go

```diff
@@ -239,3 +265,53 @@ func concatErrors(errs []error, delimiter string) error {
 	}
 	return errors.New(message)
 }
+
+// We copy lineReader, guessJSONStream, hasJSONPrefix, jsonPrefix, and hasPrefix
+// from https://github.com/kubernetes/apimachinery/blob/1da46c3f5a5b4a0cc756cb6050df0cf6f06b64c2/pkg/util/yaml/decoder.go#L347
+// because lineReader does not have a public constructor and so that we can
+// refine the return type of guessJSONStream from *io.Reader to *bufio.Reader.
+type lineReader struct {
+	reader *bufio.Reader
+}
```

Comment:
From what I can see, this reader only admits a single-line json, which is very limiting. Instead, how about assuming first the input is json, try unmarhalling, and if that fails, fallback to using the yaml reader?

Reply:
The only reason we needed to be able to handle json input here at all is because we perform an uninject step before injecting and the uninject was producing json output when `-o json` was specified (which meant that we needed to handle json input during inject).

I've updated this so that the uninject that happens as a first step to inject will always produce yaml as its (intermediate) output so that we don't need to add json input support at all.